pkg/api/validate/dynamic/sku.go

```diff
@@ -0,0 +1,71 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/computeskus"
+)
+
+// ValidateWorkerSku uses resourceSkusClient to ensure that the VM sizes listed in the cluster document are available for use in the target region.
+func (dv *dynamic) ValidateVmSku(ctx context.Context, location string, subscriptionID string, oc *api.OpenShiftCluster) error {
+	// Get a list of available worker SKUs, filtering by location.
+	filter := fmt.Sprintf("location eq %s", location)
+	skus, err := dv.resourceSkusClient.List(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	filteredSkus := computeskus.FilterVmSizes(skus, location)
+
+	// Ensure sku in MasterProfile exists in target region
+	masterSkuAvailable := false
+	masterProfileSku := string(oc.Properties.MasterProfile.VMSize)
+
+	if filteredSkus[masterProfileSku] == nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", "The selected SKU '%v' is unavailable in region '%v'", masterProfileSku, location)
+	} else {
+		masterSkuAvailable = true
+	}
+
+	// Fail if sku is available, but restricted within the subscription. Restrictions are subscription-specific.
+	// https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/error-sku-not-available
+	for _, restriction := range *filteredSkus[masterProfileSku].Restrictions {
+		restrictedLocation := strings.Join(*restriction.RestrictionInfo.Locations, "")
+
+		if restrictedLocation == location &&
+			masterSkuAvailable {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", "The selected SKU '%v' is restricted in region '%v' for selected subscription", masterProfileSku, location)
+		}
+	}
```

Comment:
This piece of code is exactly the same as the piece of code below inside of the `for` loop. Let's reduce code duplication.

Reply:
Restriction checks for masters should work in the exactly the same way as checks for workers. The only difference is what we return in the error message (`properties.masterProfile` vs `properties.workerProfiles[%d]`). I think this check should be made in a separate function.

So it will be something like this:

```go
err := validateVMRestrictions(filteredSkus, location, oc.Properties.MasterProfile.VMSize)
if err != nil {
	return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", err.Error())
}

for i, restriction := range *filteredSkus[masterProfileSku].Restrictions {
    err := validateVMRestrictions(filteredSkus, location, workerprofile.VMSize)
    if err != nil {
        return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, fm.Sprintf("properties.workerProfiles[%d].VMSize", i), err.Error())
    }
}
```

Reply:
BTW, this also makes me think that `validateVMRestrictions` should probably be placed in `pkg/util/computeskus/computeskus.go` where we already have a similar functions like `HasCapability`.

Reply:
@petrkotas @m1kola I've tried to address all comments above - the restriction check has been moved to computeskus with additional unit testing, and unnecessary code has been removed (thanks @s-amann)